compiler/noirc_evaluator/src/ssa/opt/pure.rs

```diff
@@ -512,4 +517,133 @@ mod test {
+        assert_eq!(purities[&FunctionId::test_new(0)], Purity::Impure);
+        // Impure due to mutual recursion fallback.
+        assert_eq!(purities[&FunctionId::test_new(1)], Purity::Impure);
+        assert_eq!(purities[&FunctionId::test_new(2)], Purity::Impure);
+    }
+
+    /// TODO(https://github.com/noir-lang/noir/issues/9444)
+    #[test]
+    fn brillig_functions_never_pure() {
+        let src = "
+        brillig(inline) fn main f0 {
+          b0(v0: u1):
+            call f1()
+            call f1()
+            return
+        }
+        brillig(inline) fn pure_basic f1 {
+          b0():
+            v2 = make_array [Field 0, Field 1] : [Field; 2]
+            v4 = array_get v2, index u32 1 -> Field
+            v5 = allocate -> &mut Field
+            store Field 0 at v5
+            return
+        }
+        ";
+
+        let ssa = Ssa::from_str(src).unwrap();
+        let ssa = ssa.purity_analysis();
+
+        let purities = &ssa.main().dfg.function_purities;
+        // PureWithPredicates is the default purity for all Brillig functions.
+        // So even though `f1` is technically pure it will be marked as PureWithPredicates
+        assert_eq!(purities[&FunctionId::test_new(0)], Purity::PureWithPredicate);
+        assert_eq!(purities[&FunctionId::test_new(1)], Purity::PureWithPredicate);
+    }
+
+    #[test]
+    fn call_to_function_value() {
+        let src = r#"
+        acir(inline) fn main f0 {
+          b0(v0: u32):
+            v5 = make_array [f1, f2] : [function; 2]
+            v7 = lt v0, u32 2
+            constrain v7 == u1 1, "Index out of bounds"
+            v9 = array_get v5, index v0 -> function
+            call v9()
+            return
+        }
+        acir(inline) fn lambda f1 {
+          b0():
+            return
+        }
+        acir(inline) fn lambda f2 {
+          b0():
+            return
+        }"#;
+
+        let ssa = Ssa::from_str(src).unwrap();
+        let ssa = ssa.purity_analysis();
+
+        let purities = &ssa.main().dfg.function_purities;
+        // Even though the functions referenced by the function values are pure
+        // we assume the worse case for functions containing calls to function values.
+        assert_eq!(purities[&FunctionId::test_new(0)], Purity::Impure);
+        assert_eq!(purities[&FunctionId::test_new(1)], Purity::Pure);
+        assert_eq!(purities[&FunctionId::test_new(1)], Purity::Pure);
+
+        assert_ssa_snapshot!(ssa, @r#"
+        acir(inline) impure fn main f0 {
+          b0(v0: u32):
+            v3 = make_array [f1, f2] : [function; 2]
+            v5 = lt v0, u32 2
+            constrain v5 == u1 1, "Index out of bounds"
+            v7 = array_get v3, index v0 -> function
+            call v7()
+            return
+        }
+        acir(inline) pure fn lambda f1 {
+          b0():
+            return
+        }
+        acir(inline) pure fn lambda f2 {
+          b0():
+            return
+        }
+        "#);
+    }
```

Comment:
Can probably remove this as the purity assertions above are the interesting thing.